File: 46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python3
#
# [46] 全排列
#

# @lc code=start
class Solution:
    def __init__(self):
        self.index = 0
    
    def permute(self, nums: List[int]) -> List[List[int]]:
        length1 = len(nums)
        all_range_num = self.factorial(length1)
        res = [[0 for _ in range(length1)] for _ in range(all_range_num)]
        
        def backTrack(temp_nums, path, length):
            if length == 0:
                for i in range(length1):
                    res[self.index][i] = path[i]
                self.index += 1
                return
            for i in range(length):
                curr = temp_nums[i]
                path.append(curr)
                temp_nums = temp_nums[:i] + temp_nums[i+1:]
                backTrack(temp_nums, path, len(temp_nums))
                temp_nums = temp_nums[:i] + [curr] + temp_nums[i:]
                path.pop()
        
        backTrack(nums, [], length1)
        return res
    
    def factorial(self, num):
        res = 1
        while num != 0:
            res *= num
            num -= 1
        return res
        

# @lc code=end

File: 46.全排列.go
/*
 * @lc app=leetcode.cn id=46 lang=golang
 *
 * [46] 全排列
 */

// @lc code=start
func permute(nums []int) [][]int {
	oneLength := len(nums)
	allRanges := factorial(oneLength)
	res := make([][]int, allRanges)
	index := 0
	for i := 0; i < allRanges; i++ {
		res[i] = make([]int, oneLength)
	}

	var backTrack func(tempNums, path []int, length int) 
	backTrack = func(tempNums, path []int, length int) {
		if length == 0 {
			for i := 0; i < oneLength; i++ {
				res[index][i] = path[i]
			}
			index++
			return
		}

		for i := 0; i < length; i++ {
			curr := tempNums[i]
			path = append(path, curr)
			tempNums = append(tempNums[:i], tempNums[i+1:]...)
			backTrack(tempNums, path, len(tempNums))
			tempNums = append(tempNums[:i], append([]int{curr}, tempNums[i:]...)...)
			path = path[:len(path) - 1]
		}
	}
	backTrack(nums, []int{}, oneLength)
	return res
}

func factorial(num int) int {
	res := 1
	for num != 0 {
		res *= num
		num--
	}
	return res
}
// @lc code=end

